Added equipment tag to allow specialization of equipment

AndreasBrostrom opened the pull request:
**When merged this pull request will:**
- Added a equipment tag system 
    - System allow specialized equipment per classname
    - The system have auto category system as well
       globaly (`common`), side (`common<side>`), company (`common<side>Company<company>`) and platoon (`Common<side>Company<company><platoon>`).
    - The system allow for tags refereing tags allow for item collections t o be referenced.
- Moved  arsenal under its own function folder for easier handling

> [!NOTE]
> This contians changes to the `description.ext´
> It contains a addition of a include:
> ```sqf
> #include "cScripts\CfgEquipmentTags.hpp
> ```

pool011 commented on cScripts/Loadouts/CfgLoadouts_Common.hpp:
```diff
@@ -15,7 +15,7 @@ class CommonBlufor {
     //postLoadout = "";
 
-    //role = "";
+    //equipmentTags[] = {""role_;XXxENDLINEXXx
```
> role_;XXxENDLINEXXx 

Unsure what this is

AndreasBrostrom replied:
Haha ups! my find replace did not get translated propperly xD

pool011 commented on cScripts/Loadouts/CfgLoadouts_Bravo_Atlas.hpp:
```diff
@@ -4,7 +4,8 @@ class Cav_B_Bravo_Atlas_base_F: Cav_B_Bravo_base_F {
     abilityMedic = 2;
     
-    role = "doctor";
+    equipmentTags[] = {"role_doctor"};
```
```suggestion
    equipmentTags[] = {"tag_doctor"};
```

pool011 commented on cScripts/Loadouts/CfgLoadouts_Bravo_Viking.hpp:
```diff
@@ -30,7 +30,7 @@ class Cav_B_B_Scout_PlatoonMedic_F: Cav_B_B_Scout_Officer_F {
     insignia = "cav_insignia_specialized_cls";
     abilityMedic = 2;
-    role = "medic";
+    equipmentTags[] = {"role_medic"};
```
```suggestion
    equipmentTags[] = {"tag_medic"};
```

pool011 commented on cScripts/Loadouts/CfgLoadouts_Bravo_Viking.hpp:
```diff
@@ -110,7 +110,7 @@ class Cav_B_B_Scout_CombatLifeSaver_F: Cav_B_B_Scout_Base_F {
     insignia = "cav_insignia_specialized_cls";
     abilityMedic = 1;
-    role = "medic";
+    equipmentTags[] = {"tag_medic"};
```
```suggestion
    equipmentTags[] = {"tag_cls"};
```

pool011 commented on cScripts/Loadouts/CfgLoadouts_Charlie_Squad.hpp:
```diff
@@ -26,7 +26,7 @@ class Cav_B_C_PlatoonMedic_F: Cav_B_C_Officer_F {
     insignia = "cav_insignia_specialized_cls";
     abilityMedic = 2;
-    role = "medic";
+    equipmentTags[] = {"role_medic"};
```
```suggestion
    equipmentTags[] = {"tag_medic"};
```

pool011 reviewed (changes requested):
Have to straighten out the first few roles we have added. Then we can work on adding the "sidearm" tag for people who are authorized to carry a sidearm but don't have one equipped by default.

pool011 commented on cScripts/functions/systems/fn_getArsenalWhitelist.sqf:
```diff
@@ -37,26 +37,26 @@ private _companyItems = switch (_company) do {
+        case "role_pilotfighter": {GET_CONTAINER_KEYS("arsenal_role_pilotfighter");};
+        case "role_pilottransport": {GET_CONTAINER_KEYS("arsenal_role_pilottransport");};
+        
+        case "tag_cls": {GET_CONTAINER_KEYS("arsenal_role_cls");};
+        case "tag_medic": {GET_CONTAINER_KEYS("arsenal_role_medic");};
+        case "tag_doctor": {GET_CONTAINER_KEYS("arsenal_role_doctor");};
```
```suggestion
        case "tag_cls": {GET_CONTAINER_KEYS("arsenal_tag_cls");};
        case "tag_medic": {GET_CONTAINER_KEYS("arsenal_tag_medic");};
        case "tag_doctor": {GET_CONTAINER_KEYS("arsenal_tag_doctor");};
```

I would go into `init_logistics.sqf` and change these container names to match the suffixes we are looking for. in the switch here to maintain a consistent naming convention.

pool011 reviewed (commented):
Naming convention changes should be made for the new tags

AndreasBrostrom commented:
# Wiki Article: [Gear.md](https://github.com/7Cav/cScripts/files/14896644/Gear.md)
cScripts Gear are our loadout script strongly inspired and taken from [Poppy](https://github.com/BaerMitUmlaut/Poppy/), our old loadout script. The gear system is rewritten and is greatly tailored to our community needs and have a extra simplification added to it. The loadout utilized a loadout array obtained from ace export loadout or the bis [`getUnitLoadout`](https://community.bistudio.com/wiki/getUnitLoadout) script command. The config also for customizing and can handle abilities and unit cosmetics such as rank and insignia.

### Loadouts
Loadouts are automatically applied to units on spawn based on the variable, variableName, classname or side. Loadouts are also saved when altered (*See below*). Loadouts using `Public` scope (*[See below](https://github.com/7Cav/cScripts/wiki/Gear#how-to-gear-up)*) in the config will also be present in loadout selectors such as ACE Arsenal default loadouts and [[staging]] and [[starter crate]] selectors.

### Loadout Saving
When you alter your loadout it automatically get saved and reapplied on respawn. The save event is run when you alter the loadout in **arsenal** or when you **close your inventory** and only happens when your inside of a defined [[Staging Zone|staging zone]]. All gear related functions will honer the saved loadout. Permissions are also kept based on the core loadout used before the save was committed.

## How to gear up
The gear system is quite simple it utilized config values to obtain and apply your loadout.
Loadouts are defined in the `CfgLoadouts.hpp`. The system allow loadout application for any side and unit and are inherited from parent classnames.
In order to make loadout for a side can setup a `Common<Blufor|Opfor|Independant|Civilian>` baseclass an example for Bluefor side can be found below.

```cpp
class CommonBlufor {};
class My_Loadout : CommonBlufor {
    regiment = "";
    company = "";
    platoon = -1;
    squad = -1;
    team = "";

    displayName = "";
    scope = 0;
    category[] = {};
    loadout = [[],[],[],[],[],[],"","",[],["","","","","",""]];
    equipmentTags[] = {};

    icon = "";
    insignia = "";

    abilityMedic = 0;
    abilityEngineer = 0;
    abilityEOD = 0;

    preLoadout = "";
    postLoadout = "";
};
```
### Config Properties
```cpp
class CommonBlufor {};
class My_Loadout : CommonBlufor {
CommonBlufor {
    regiment = "";          // Style value
    company = "";           // Used for arsenal whitelist based on side
    platoon = -1;           // Style value
    squad = -1;             // Used to determine squad name
    team = "";              // Used for auto coloring of fire team

    displayName = "";       // Used to make the selection label
    scope = 0;              // 0 private: not obtainable. 1 protected: can be used but not selected. 2 public: can be selected and used
    category[] = {};        // Hardcoded category paths found in ".\cScripts\CavFnc\functions\systems\fn_setupLoadoutCategories.sqf"
    loadout = [[],[],[],[],[],[],"","",[],["","","","","",""]]; // loadout array none quoted warped
    equipmentTags[] = {};   // Equipment tags

    icon = "";              // Style value (use classname of a unit icon)
    insignia = "";          // Style value (use classname of a insignia)

    abilityMedic = 0;       // 0 Nothing, 1 Medic, 2 Doctor
    abilityEngineer = 0;    // 0 Nothing, 1 Repair Specialists, 2 Engineer
    abilityEOD = 0;         // 0 Nothing, 1 EOD Specialist

    preLoadout = "";        // code that gets applied befor loadout and abilities are applied
    postLoadout = "";       // code that gets applied after loadout and abilities are applied
};
```

## Loadout applications
<img align="right" width="300" height="105" src="https://github.com/7Cav/cScripts/blob/main/resourses/wikigfx/gear_applyloadout_examples.png">Loadouts can be applied in several ways. The main system used is the **classname** based system but it can also apply to units using **variableName**, **setVariable** or **side**. The loadout also need to be `public` or `protected` scope to be automatically applied, `public` to be select it via a selector.
```cpp
this setVariable ["cScripts_Gear_LoadoutClass", "Cav_B_C_Rifleman_F"];
```
### Keep spawn Loadout
You can keep the EDEN selected loadout by adding the following variable to a player unit:
```cpp
this setVariable ["cScripts_Gear_LoadoutClass", "default"];
```
_In our current setup of our loadouts `default` is not defined. This means it will not be applied. A warning of this will be shown in Eden SP and Eden MP._

## Arsenal
The arsenal is built automatically when opening the staging variant of the arsenal and utilizes side, company, weapon system and equipmentTags to construct the whitelist.

### Equipment Tags 
Equipment tags are defined in `CfgEquipmentTags.hpp` and called from a loadout in the `equipmentTags[]` array in `CfgLoadouts.hpp`. The equipment tags similar to the loadout system have side bases tags `Common<Blufor|Opfor|Independant|Civilian>` as well for company. Companies are defined as `Common<Blufor|Opfor|Independant|Civilian>Company<CompanyName>`. It is also possible to decare your own equipmentTags by adding a new config entry.
There are also one global equipment tag that is always applied to all whitelists named `Common`.

```
    Common[] = {"G_Aviator"};
    CommonBlufor[] = {
        "rhs_weap_mk18_KAC_bk",
        "rhs_weap_mk18_KAC",
        "rhs_mag_30Rnd_556x45_M855A1_PMAG_Tracer_Red"
    };
    CommonBluforCompanyCharlie[] = {"ace_banana"};
```

## See Also
- [[Staging Zone]]
- [[Starter Crate]]
-

AndreasBrostrom committed:
Addeed equipment tag system

AndreasBrostrom committed:
Adjusted Tag list

AndreasBrostrom committed:
Renamed the arsenal system to be prefixed and saved in one category
Added addArsenal Wrapper to be used for missions were you want a action to be added.

AndreasBrostrom committed:
Fixed missing closing bracket

AndreasBrostrom committed:
Fixed item tag being able to add tags

AndreasBrostrom committed:
Added error handling

AndreasBrostrom committed:
Type type check

AndreasBrostrom committed:
Added filtered item function

AndreasBrostrom committed:
aresnal removed in this version

AndreasBrostrom committed:
Updated handling of Organization and implemented OrgArray

AndreasBrostrom committed:
Added a item filter system and fixed syntax

AndreasBrostrom committed:
Fixed space in macro

AndreasBrostrom committed:
improved endles loop detection

AndreasBrostrom committed:
Update fn_gear_getTagItems.sqf

AndreasBrostrom committed:
Update fn_arsenal_clearDefaultLoadouts.sqf

AndreasBrostrom committed:
Merge branch 'main' into EquipmentTag
